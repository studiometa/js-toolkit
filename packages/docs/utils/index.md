# Utils

Find below functions that can help you in your projects.

## Misc.

- [debounce](/utils/debounce.html)
- [focusTrap](/utils/focusTrap.html)
- [keyCodes](/utils/keyCodes.html)
- [memoize](/utils/memoize.html)
- [nextFrame](/utils/nextFrame.html)
- [scrollTo](/utils/scrollTo.html)
- [throttle](/utils/throttle.html)

## Collision

- [boundingRectToCircle](/utils/collision/boundingRectToCircle.html)
- [collideCircleCircle](/utils/collision/collideCircleCircle.html)
- [collideCircleRect](/utils/collision/collideCircleRect.html)
- [collidePointCircle](/utils/collision/collidePointCircle.html)
- [collidePointRect](/utils/collision/collidePointRect.html)
- [collideRectRect](/utils/collision/collideRectRect.html)

## CSS

- [matrix](/utils/css/matrix.html)
- [transition](/utils/css/transition.html)

## History

- [historyPush](/utils/history/historyPush.html)
- [historyReplace](/utils/history/historyReplace.html)
- [objectToURLSearchParams](/utils/history/objectToURLSearchParams.html)

## Math

- [clamp](/utils/math/clamp.html)
- [clamp01](/utils/math/clamp01.html)
- [damp](/utils/math/damp.html)
- [inertiaFinalValue](/utils/math/inertiaFinalValue.html)
- [lerp](/utils/math/lerp.html)
- [map](/utils/math/map.html)
- [round](/utils/math/round.html)

## String

- [withLeadingCharacters](/utils/string/withLeadingCharacters.html)
- [withLeadingSlash](/utils/string/withLeadingSlash.html)
- [withoutLeadingCharacters](/utils/string/withoutLeadingCharacters.html)
- [withoutLeadingCharactersRecursive](/utils/string/withoutLeadingCharactersRecursive.html)
- [withoutLeadingSlash](/utils/string/withoutLeadingSlash.html)
- [withoutTrailingCharacters](/utils/string/withoutTrailingCharacters.html)
- [withoutTrailingCharactersRecursive](/utils/string/withoutTrailingCharactersRecursive.html)
- [withoutTrailingSlash](/utils/string/withoutTrailingSlash.html)
- [withTrailingCharacters](/utils/string/withTrailingCharacters.html)
- [withTrailingSlash](/utils/string/withTrailingSlash.html)
