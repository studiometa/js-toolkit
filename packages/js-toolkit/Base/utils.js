import { isArray, isDefined, isDev } from '../utils/index.js';

/**
 * @typedef {import('./index.js').default} Base
 */

/**
 * Get a list of elements based on the name of a component.
 *
 * @param {string} nameOrSelector
 *   The name or selector to used for this component.
 * @param {HTMLElement|Document} element
 *   The root element on which to query the selector, defaults to `document`.
 * @returns {Array<HTMLElement>}
 *   A list of elements on which the component should be mounted.
 */
export function getComponentElements(nameOrSelector, element = document) {
  const selector = `[data-component="${nameOrSelector}"]`;
  let elements = [];

  try {
    elements = Array.from(element.querySelectorAll(selector));
    // eslint-disable-next-line no-empty
  } catch {}

  // If no child component found with the default selector, try a classic DOM selector
  if (elements.length === 0) {
    elements = Array.from(element.querySelectorAll(nameOrSelector));
  }

  return elements;
}

/**
 * Test if an event is defined in the given config.
 *
 * @param   {string} event
 * @param   {import('./index.js').BaseConfig} config
 * @returns {boolean}
 */
export function eventIsDefinedInConfig(event, config) {
  return isArray(config.emits) && config.emits.includes(event);
}

/**
 * Test if an event can be used on the given element.
 *
 * @param   {string} event
 * @param   {HTMLElement} element
 * @returns {boolean}
 */
export function eventIsNative(event, element) {
  return isDefined(element[`on${event}`]);
}

/**
 * Get the target of a given event.
 *
 * @param   {Base} instance
 * @param   {string} event
 * @param   {import('./index.js').BaseConfig} config
 * @returns {Base|Base['$el']}
 */
export function getEventTarget(instance, event, config) {
  if (eventIsDefinedInConfig(event, config)) {
    return instance;
  }

  if (eventIsNative(event, instance.$el)) {
    return instance.$el;
  }

  if (isDev) {
    console.warn(
      `[${config.name}]`,
      `The "${event}" event is missing from the configuration and is not a native`,
      `event for the root element of type \`${instance.$el.constructor.name}\`.`
    );
  }

  return instance;
}
