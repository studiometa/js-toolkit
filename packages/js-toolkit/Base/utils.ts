import { isArray, isDefined, isDev, SmartQueue } from '../utils/index.js';
import type { Base, BaseConfig } from './index.js';
import { features } from './features.js';

let queue: SmartQueue;

/**
 * Add a task to the main queue.
 */
export function addToQueue(fn: () => unknown): void {
  if (!features.get('asyncChildren')) {
    fn();
    return;
  }

  if (!queue) {
    queue = new SmartQueue();
  }

  queue.add(fn);
}

const selectors = new Map();

// Separator used for multi-component declaration in `data-component` attributeS.
const separator = ' ';

/**
 * Get the selector for a given component.
 */
function getSelector(nameOrSelector: string): string {
  if (!selectors.has(nameOrSelector)) {
    const parts = [
      // Single selector
      `[data-component="${nameOrSelector}"]`,
      // Selector in the middle of a list of selectors
      `[data-component*="${separator}${nameOrSelector}${separator}"]`,
      // Selector at the end of a list of selectors
      `[data-component$="${separator}${nameOrSelector}"]`,
      // Selector at the beginning of a list of selectors
      `[data-component^="${nameOrSelector}${separator}"]`,
    ];
    selectors.set(nameOrSelector, parts.join(','));
  }

  return selectors.get(nameOrSelector);
}

/**
 * Get a list of elements based on the name of a component.
 *
 * @param {string} nameOrSelector
 *   The name or selector to used for this component.
 * @param {HTMLElement|Document} element
 *   The root element on which to query the selector, defaults to `document`.
 * @returns {Array<HTMLElement>}
 *   A list of elements on which the component should be mounted.
 */
export function getComponentElements(
  nameOrSelector: string,
  element: HTMLElement | Document = document,
): HTMLElement[] {
  const selector = getSelector(nameOrSelector);
  let elements = [];

  try {
    elements = Array.from(element.querySelectorAll(selector));
    // eslint-disable-next-line no-empty
  } catch {}

  // If no child component found with the default selector, try a classic DOM selector
  if (elements.length === 0) {
    elements = Array.from(element.querySelectorAll(nameOrSelector));
  }

  return elements;
}

/**
 * Test if an event is defined in the given config.
 */
export function eventIsDefinedInConfig(event: string, config: BaseConfig): boolean {
  return isArray(config.emits) && config.emits.includes(event);
}

/**
 * Test if an event can be used on the given element.
 */
export function eventIsNative(event: string, element: HTMLElement): boolean {
  return isDefined(element[`on${event}`]);
}

/**
 * Get the target of a given event.
 * @todo Return false in v3 if event is not defined or not native to prevent adding a listener.
 */
export function getEventTarget(
  instance: Base,
  event: string,
  config: BaseConfig,
): Base | Base['$el'] {
  if (eventIsDefinedInConfig(event, config)) {
    return instance;
  }

  if (eventIsNative(event, instance.$el)) {
    return instance.$el;
  }

  if (isDev) {
    console.warn(
      `[${config.name}]`,
      `The "${event}" event is missing from the configuration and is not a native`,
      `event for the root element of type \`${instance.$el.constructor.name}\`.`,
    );
  }

  // @todo v3 return false or null
  return instance;
}
