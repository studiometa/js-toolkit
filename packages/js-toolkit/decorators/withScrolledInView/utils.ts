import { isNumber, isString } from '../../utils/index.js';
import type { NormalizedOffset } from './types.js';

const units = {
  px: 'px',
  vh: 'vh',
  vw: 'vw',
  vmin: 'vmin',
  vmax: 'vmax',
  percent: '%',
};

const namedOffsets = {
  start: 0,
  center: 0.5,
  end: 1,
};

/**
 * Normalize offset input.
 */
export function normalizeOffset(offsets: string): NormalizedOffset {
  return offsets
    .split('/')
    .map((offset) => offset.trim().split(' ').slice(0, 2)) as NormalizedOffset;
}

/**
 * Parse the given offset.
 */
export function parseNamedOffset(offset: string | number): string | number {
  if (isNumber(offset)) {
    return offset;
  }

  if (isNumber(namedOffsets[offset])) {
    return namedOffsets[offset];
  }

  if (isNumber(Number(offset))) {
    return Number(offset);
  }

  if (isString(offset) && offset.endsWith(units.percent)) {
    return Number.parseFloat(offset) / 100;
  }

  return offset;
}

/**
 * Get an edge value given a start, size and offset value.
 */
export function getEdgeWithOffset(start: number, size: number, offset: string | number): number {
  if (isNumber(offset)) {
    return start + size * offset;
  }

  const parsedNamedOffset = parseNamedOffset(offset);
  if (isNumber(parsedNamedOffset)) {
    return start + size * parsedNamedOffset;
  }

  if (isString(parsedNamedOffset)) {
    const parsedOffset = Number.parseFloat(parsedNamedOffset);

    if (parsedNamedOffset.endsWith(units.px)) {
      return start + parsedOffset;
    }

    if (parsedNamedOffset.endsWith(units.vh)) {
      return start + (parsedOffset * window.innerHeight) / 100;
    }

    if (parsedNamedOffset.endsWith(units.vw)) {
      return start + (parsedOffset * window.innerWidth) / 100;
    }

    if (parsedNamedOffset.endsWith(units.vmin)) {
      return start + (parsedOffset * Math.min(window.innerWidth, window.innerHeight)) / 100;
    }

    if (parsedNamedOffset.endsWith(units.vmax)) {
      return start + (parsedOffset * Math.max(window.innerWidth, window.innerHeight)) / 100;
    }
  }

  return start;
}

/**
 * Get starting and ending edges for a given axis, a target sizings, a container sizings and their offset.
 */
export function getEdges(axis: 'x' | 'y', targetSizes, containerSizes, offset): [number, number] {
  const sizeKey = axis === 'x' ? 'width' : 'height';
  const targetStart = getEdgeWithOffset(targetSizes[axis], targetSizes[sizeKey], offset[0][0]);
  const containerStart = getEdgeWithOffset(0, containerSizes[sizeKey], offset[0][1]);
  const targetEnd = getEdgeWithOffset(targetSizes[axis], targetSizes[sizeKey], offset[1][0]);
  const containerEnd = getEdgeWithOffset(0, containerSizes[sizeKey], offset[1][1]);
  const start = targetStart - containerStart;
  const end = targetEnd - containerEnd;

  return [start, end];
}
