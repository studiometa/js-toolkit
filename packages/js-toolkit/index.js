export { default as Base } from './Base/index.js';
export * from './decorators/index.js';
export * from './helpers/index.js';
export * from './services/index.js';
