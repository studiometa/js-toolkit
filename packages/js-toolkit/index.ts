export { Base } from './Base/index.js';
export * from './decorators/index.js';
export * from './helpers/index.js';
export * from './services/index.js';

export type {
  BaseTypeParameter,
  BaseConstructor,
  BaseAsyncConstructor,
  BaseConfig,
} from './Base/index.js';

export type { BaseInterface, BaseDecorator } from './Base/types.js';
