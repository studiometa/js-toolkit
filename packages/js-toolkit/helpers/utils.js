/* eslint-disable import/prefer-default-export */
import { getComponentElements } from '../Base/utils.js';

/**
 * Get the target elements for the lazy import helper functions.
 *
 * @param {string|HTMLElement|HTMLElement[]} nameOrSelectorOrElement
 *   The original selector or element, or list of elements.
 * @param {HTMLElement} [context]
 *   The optional context to use to query for elements.
 * @returns {HTMLElement[]} A normalized list of elements.
 */
export function getTargetElements(nameOrSelectorOrElement, context) {
  if (typeof nameOrSelectorOrElement === 'string') {
    return getComponentElements(nameOrSelectorOrElement, context);
  }

  if (!Array.isArray(nameOrSelectorOrElement)) {
    return [nameOrSelectorOrElement];
  }

  return nameOrSelectorOrElement;
}
