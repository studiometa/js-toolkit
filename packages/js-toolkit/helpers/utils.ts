/* eslint-disable import/prefer-default-export */
import { getComponentElements } from '../Base/utils.js';
import { isString, isArray } from '../utils/index.js';

/**
 * Get the target elements for the lazy import helper functions.
 *
 * @param {string|HTMLElement|HTMLElement[]} nameOrSelectorOrElement
 *   The original selector or element, or list of elements.
 * @param {HTMLElement} [context]
 *   The optional context to use to query for elements.
 * @return {HTMLElement[]} A normalized list of elements.
 */
export function getTargetElements(
  nameOrSelectorOrElement: string | HTMLElement | HTMLElement[],
  context: HTMLElement,
): HTMLElement[] {
  if (isString(nameOrSelectorOrElement)) {
    return getComponentElements(nameOrSelectorOrElement, context);
  }

  if (!isArray(nameOrSelectorOrElement)) {
    return [nameOrSelectorOrElement];
  }

  return nameOrSelectorOrElement;
}
