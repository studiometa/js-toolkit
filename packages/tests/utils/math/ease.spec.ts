import { describe, it, expect } from 'bun:test';
import { ease } from '@studiometa/js-toolkit/utils';

const values = Array(10)
  .fill(0)
  .map((value, index, array) => index / (array.length - 1));

describe('ease methods', () => {
  Object.keys(ease)
    .filter((key) => key.startsWith('ease'))
    .forEach((key) => {
      it(`the ${key} method should give the correct value`, () => {
        values.forEach((value) => {
          expect(ease[key](value)).toMatchSnapshot();
        });
      });
    });
});
