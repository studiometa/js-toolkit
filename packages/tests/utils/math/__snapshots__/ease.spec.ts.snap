// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ease methods the easeInCirc method should give the correct value 1`] = `-0`;

exports[`ease methods the easeInCirc method should give the correct value 2`] = `0.006192010000093506`;

exports[`ease methods the easeInCirc method should give the correct value 3`] = `0.025003956956430873`;

exports[`ease methods the easeInCirc method should give the correct value 4`] = `0.057190958417936644`;

exports[`ease methods the easeInCirc method should give the correct value 5`] = `0.10419358352238339`;

exports[`ease methods the easeInCirc method should give the correct value 6`] = `0.1685205807169019`;

exports[`ease methods the easeInCirc method should give the correct value 7`] = `0.2546440075000701`;

exports[`ease methods the easeInCirc method should give the correct value 8`] = `0.37146063894529113`;

exports[`ease methods the easeInCirc method should give the correct value 9`] = `0.5418771527091488`;

exports[`ease methods the easeInCirc method should give the correct value 10`] = `1`;

exports[`ease methods the easeInCubic method should give the correct value 1`] = `0`;

exports[`ease methods the easeInCubic method should give the correct value 2`] = `0.0013717421124828531`;

exports[`ease methods the easeInCubic method should give the correct value 3`] = `0.010973936899862825`;

exports[`ease methods the easeInCubic method should give the correct value 4`] = `0.037037037037037035`;

exports[`ease methods the easeInCubic method should give the correct value 5`] = `0.0877914951989026`;

exports[`ease methods the easeInCubic method should give the correct value 6`] = `0.1714677640603567`;

exports[`ease methods the easeInCubic method should give the correct value 7`] = `0.2962962962962963`;

exports[`ease methods the easeInCubic method should give the correct value 8`] = `0.4705075445816187`;

exports[`ease methods the easeInCubic method should give the correct value 9`] = `0.7023319615912208`;

exports[`ease methods the easeInCubic method should give the correct value 10`] = `1`;

exports[`ease methods the easeInExpo method should give the correct value 1`] = `0`;

exports[`ease methods the easeInExpo method should give the correct value 2`] = `0.002109491677524035`;

exports[`ease methods the easeInExpo method should give the correct value 3`] = `0.004556754060844206`;

exports[`ease methods the easeInExpo method should give the correct value 4`] = `0.009843133202303688`;

exports[`ease methods the easeInExpo method should give the correct value 5`] = `0.021262343752724643`;

exports[`ease methods the easeInExpo method should give the correct value 6`] = `0.045929202883612456`;

exports[`ease methods the easeInExpo method should give the correct value 7`] = `0.09921256574801243`;

exports[`ease methods the easeInExpo method should give the correct value 8`] = `0.2143109957132682`;

exports[`ease methods the easeInExpo method should give the correct value 9`] = `0.46293735614364506`;

exports[`ease methods the easeInExpo method should give the correct value 10`] = `1`;

exports[`ease methods the easeInOutCirc method should give the correct value 1`] = `0`;

exports[`ease methods the easeInOutCirc method should give the correct value 2`] = `0.012501978478215436`;

exports[`ease methods the easeInOutCirc method should give the correct value 3`] = `0.052096791761191696`;

exports[`ease methods the easeInOutCirc method should give the correct value 4`] = `0.12732200375003505`;

exports[`ease methods the easeInOutCirc method should give the correct value 5`] = `0.2709385763545744`;

exports[`ease methods the easeInOutCirc method should give the correct value 6`] = `0.7290614236454256`;

exports[`ease methods the easeInOutCirc method should give the correct value 7`] = `0.8726779962499649`;

exports[`ease methods the easeInOutCirc method should give the correct value 8`] = `0.9479032082388084`;

exports[`ease methods the easeInOutCirc method should give the correct value 9`] = `0.9874980215217846`;

exports[`ease methods the easeInOutCirc method should give the correct value 10`] = `1`;

exports[`ease methods the easeInOutCubic method should give the correct value 1`] = `0`;

exports[`ease methods the easeInOutCubic method should give the correct value 2`] = `0.0054869684499314125`;

exports[`ease methods the easeInOutCubic method should give the correct value 3`] = `0.0438957475994513`;

exports[`ease methods the easeInOutCubic method should give the correct value 4`] = `0.14814814814814814`;

exports[`ease methods the easeInOutCubic method should give the correct value 5`] = `0.3511659807956104`;

exports[`ease methods the easeInOutCubic method should give the correct value 6`] = `0.6488340192043895`;

exports[`ease methods the easeInOutCubic method should give the correct value 7`] = `0.8518518518518519`;

exports[`ease methods the easeInOutCubic method should give the correct value 8`] = `0.9561042524005487`;

exports[`ease methods the easeInOutCubic method should give the correct value 9`] = `0.9945130315500685`;

exports[`ease methods the easeInOutCubic method should give the correct value 10`] = `1`;

exports[`ease methods the easeInOutExpo method should give the correct value 1`] = `0`;

exports[`ease methods the easeInOutExpo method should give the correct value 2`] = `0.002278377030422103`;

exports[`ease methods the easeInOutExpo method should give the correct value 3`] = `0.010631171876362321`;

exports[`ease methods the easeInOutExpo method should give the correct value 4`] = `0.049606282874006216`;

exports[`ease methods the easeInOutExpo method should give the correct value 5`] = `0.23146867807182253`;

exports[`ease methods the easeInOutExpo method should give the correct value 6`] = `0.7685313219281775`;

exports[`ease methods the easeInOutExpo method should give the correct value 7`] = `0.9503937171259937`;

exports[`ease methods the easeInOutExpo method should give the correct value 8`] = `0.9893688281236377`;

exports[`ease methods the easeInOutExpo method should give the correct value 9`] = `0.9977216229695779`;

exports[`ease methods the easeInOutExpo method should give the correct value 10`] = `1`;

exports[`ease methods the easeInOutQuad method should give the correct value 1`] = `0`;

exports[`ease methods the easeInOutQuad method should give the correct value 2`] = `0.024691358024691357`;

exports[`ease methods the easeInOutQuad method should give the correct value 3`] = `0.09876543209876543`;

exports[`ease methods the easeInOutQuad method should give the correct value 4`] = `0.2222222222222222`;

exports[`ease methods the easeInOutQuad method should give the correct value 5`] = `0.3950617283950617`;

exports[`ease methods the easeInOutQuad method should give the correct value 6`] = `0.6049382716049383`;

exports[`ease methods the easeInOutQuad method should give the correct value 7`] = `0.7777777777777777`;

exports[`ease methods the easeInOutQuad method should give the correct value 8`] = `0.9012345679012346`;

exports[`ease methods the easeInOutQuad method should give the correct value 9`] = `0.9753086419753086`;

exports[`ease methods the easeInOutQuad method should give the correct value 10`] = `1`;

exports[`ease methods the easeInOutQuart method should give the correct value 1`] = `0`;

exports[`ease methods the easeInOutQuart method should give the correct value 2`] = `0.0012193263222069805`;

exports[`ease methods the easeInOutQuart method should give the correct value 3`] = `0.019509221155311687`;

exports[`ease methods the easeInOutQuart method should give the correct value 4`] = `0.09876543209876543`;

exports[`ease methods the easeInOutQuart method should give the correct value 5`] = `0.312147538484987`;

exports[`ease methods the easeInOutQuart method should give the correct value 6`] = `0.687852461515013`;

exports[`ease methods the easeInOutQuart method should give the correct value 7`] = `0.9012345679012346`;

exports[`ease methods the easeInOutQuart method should give the correct value 8`] = `0.9804907788446883`;

exports[`ease methods the easeInOutQuart method should give the correct value 9`] = `0.998780673677793`;

exports[`ease methods the easeInOutQuart method should give the correct value 10`] = `1`;

exports[`ease methods the easeInOutQuint method should give the correct value 1`] = `0`;

exports[`ease methods the easeInOutQuint method should give the correct value 2`] = `0.0002709614049348845`;

exports[`ease methods the easeInOutQuint method should give the correct value 3`] = `0.008670764957916305`;

exports[`ease methods the easeInOutQuint method should give the correct value 4`] = `0.06584362139917695`;

exports[`ease methods the easeInOutQuint method should give the correct value 5`] = `0.27746447865332174`;

exports[`ease methods the easeInOutQuint method should give the correct value 6`] = `0.7225355213466782`;

exports[`ease methods the easeInOutQuint method should give the correct value 7`] = `0.934156378600823`;

exports[`ease methods the easeInOutQuint method should give the correct value 8`] = `0.9913292350420837`;

exports[`ease methods the easeInOutQuint method should give the correct value 9`] = `0.9997290385950651`;

exports[`ease methods the easeInOutQuint method should give the correct value 10`] = `1`;

exports[`ease methods the easeInOutSine method should give the correct value 1`] = `0`;

exports[`ease methods the easeInOutSine method should give the correct value 2`] = `0.030153689607045786`;

exports[`ease methods the easeInOutSine method should give the correct value 3`] = `0.116977778440511`;

exports[`ease methods the easeInOutSine method should give the correct value 4`] = `0.24999999999999994`;

exports[`ease methods the easeInOutSine method should give the correct value 5`] = `0.4131759111665348`;

exports[`ease methods the easeInOutSine method should give the correct value 6`] = `0.5868240888334653`;

exports[`ease methods the easeInOutSine method should give the correct value 7`] = `0.75`;

exports[`ease methods the easeInOutSine method should give the correct value 8`] = `0.883022221559489`;

exports[`ease methods the easeInOutSine method should give the correct value 9`] = `0.9698463103929542`;

exports[`ease methods the easeInOutSine method should give the correct value 10`] = `1`;

exports[`ease methods the easeInQuad method should give the correct value 1`] = `0`;

exports[`ease methods the easeInQuad method should give the correct value 2`] = `0.012345679012345678`;

exports[`ease methods the easeInQuad method should give the correct value 3`] = `0.04938271604938271`;

exports[`ease methods the easeInQuad method should give the correct value 4`] = `0.1111111111111111`;

exports[`ease methods the easeInQuad method should give the correct value 5`] = `0.19753086419753085`;

exports[`ease methods the easeInQuad method should give the correct value 6`] = `0.308641975308642`;

exports[`ease methods the easeInQuad method should give the correct value 7`] = `0.4444444444444444`;

exports[`ease methods the easeInQuad method should give the correct value 8`] = `0.6049382716049383`;

exports[`ease methods the easeInQuad method should give the correct value 9`] = `0.7901234567901234`;

exports[`ease methods the easeInQuad method should give the correct value 10`] = `1`;

exports[`ease methods the easeInQuart method should give the correct value 1`] = `0`;

exports[`ease methods the easeInQuart method should give the correct value 2`] = `0.00015241579027587256`;

exports[`ease methods the easeInQuart method should give the correct value 3`] = `0.002438652644413961`;

exports[`ease methods the easeInQuart method should give the correct value 4`] = `0.012345679012345678`;

exports[`ease methods the easeInQuart method should give the correct value 5`] = `0.039018442310623375`;

exports[`ease methods the easeInQuart method should give the correct value 6`] = `0.09525986892242039`;

exports[`ease methods the easeInQuart method should give the correct value 7`] = `0.19753086419753085`;

exports[`ease methods the easeInQuart method should give the correct value 8`] = `0.3659503124523701`;

exports[`ease methods the easeInQuart method should give the correct value 9`] = `0.624295076969974`;

exports[`ease methods the easeInQuart method should give the correct value 10`] = `1`;

exports[`ease methods the easeInQuint method should give the correct value 1`] = `0`;

exports[`ease methods the easeInQuint method should give the correct value 2`] = `0.000016935087808430282`;

exports[`ease methods the easeInQuint method should give the correct value 3`] = `0.000541922809869769`;

exports[`ease methods the easeInQuint method should give the correct value 4`] = `0.004115226337448559`;

exports[`ease methods the easeInQuint method should give the correct value 5`] = `0.01734152991583261`;

exports[`ease methods the easeInQuint method should give the correct value 6`] = `0.05292214940134466`;

exports[`ease methods the easeInQuint method should give the correct value 7`] = `0.1316872427983539`;

exports[`ease methods the easeInQuint method should give the correct value 8`] = `0.28462802079628785`;

exports[`ease methods the easeInQuint method should give the correct value 9`] = `0.5549289573066435`;

exports[`ease methods the easeInQuint method should give the correct value 10`] = `1`;

exports[`ease methods the easeInSine method should give the correct value 1`] = `0`;

exports[`ease methods the easeInSine method should give the correct value 2`] = `0.01519224698779198`;

exports[`ease methods the easeInSine method should give the correct value 3`] = `0.06030737921409157`;

exports[`ease methods the easeInSine method should give the correct value 4`] = `0.1339745962155613`;

exports[`ease methods the easeInSine method should give the correct value 5`] = `0.233955556881022`;

exports[`ease methods the easeInSine method should give the correct value 6`] = `0.35721239031346064`;

exports[`ease methods the easeInSine method should give the correct value 7`] = `0.4999999999999999`;

exports[`ease methods the easeInSine method should give the correct value 8`] = `0.6579798566743311`;

exports[`ease methods the easeInSine method should give the correct value 9`] = `0.8263518223330696`;

exports[`ease methods the easeInSine method should give the correct value 10`] = `1`;

exports[`ease methods the easeOutCirc method should give the correct value 1`] = `0`;

exports[`ease methods the easeOutCirc method should give the correct value 2`] = `0.4581228472908512`;

exports[`ease methods the easeOutCirc method should give the correct value 3`] = `0.6285393610547089`;

exports[`ease methods the easeOutCirc method should give the correct value 4`] = `0.7453559924999298`;

exports[`ease methods the easeOutCirc method should give the correct value 5`] = `0.8314794192830981`;

exports[`ease methods the easeOutCirc method should give the correct value 6`] = `0.8958064164776166`;

exports[`ease methods the easeOutCirc method should give the correct value 7`] = `0.9428090415820634`;

exports[`ease methods the easeOutCirc method should give the correct value 8`] = `0.9749960430435691`;

exports[`ease methods the easeOutCirc method should give the correct value 9`] = `0.9938079899999065`;

exports[`ease methods the easeOutCirc method should give the correct value 10`] = `1`;

exports[`ease methods the easeOutCubic method should give the correct value 1`] = `0`;

exports[`ease methods the easeOutCubic method should give the correct value 2`] = `0.2976680384087792`;

exports[`ease methods the easeOutCubic method should give the correct value 3`] = `0.5294924554183813`;

exports[`ease methods the easeOutCubic method should give the correct value 4`] = `0.7037037037037036`;

exports[`ease methods the easeOutCubic method should give the correct value 5`] = `0.8285322359396433`;

exports[`ease methods the easeOutCubic method should give the correct value 6`] = `0.9122085048010974`;

exports[`ease methods the easeOutCubic method should give the correct value 7`] = `0.9629629629629629`;

exports[`ease methods the easeOutCubic method should give the correct value 8`] = `0.9890260631001372`;

exports[`ease methods the easeOutCubic method should give the correct value 9`] = `0.9986282578875172`;

exports[`ease methods the easeOutCubic method should give the correct value 10`] = `1`;

exports[`ease methods the easeOutExpo method should give the correct value 1`] = `0`;

exports[`ease methods the easeOutExpo method should give the correct value 2`] = `0.5370626438563549`;

exports[`ease methods the easeOutExpo method should give the correct value 3`] = `0.7856890042867318`;

exports[`ease methods the easeOutExpo method should give the correct value 4`] = `0.9007874342519875`;

exports[`ease methods the easeOutExpo method should give the correct value 5`] = `0.9540707971163875`;

exports[`ease methods the easeOutExpo method should give the correct value 6`] = `0.9787376562472754`;

exports[`ease methods the easeOutExpo method should give the correct value 7`] = `0.9901568667976963`;

exports[`ease methods the easeOutExpo method should give the correct value 8`] = `0.9954432459391558`;

exports[`ease methods the easeOutExpo method should give the correct value 9`] = `0.9978905083224759`;

exports[`ease methods the easeOutExpo method should give the correct value 10`] = `1`;

exports[`ease methods the easeOutQuad method should give the correct value 1`] = `0`;

exports[`ease methods the easeOutQuad method should give the correct value 2`] = `0.2098765432098766`;

exports[`ease methods the easeOutQuad method should give the correct value 3`] = `0.3950617283950617`;

exports[`ease methods the easeOutQuad method should give the correct value 4`] = `0.5555555555555555`;

exports[`ease methods the easeOutQuad method should give the correct value 5`] = `0.691358024691358`;

exports[`ease methods the easeOutQuad method should give the correct value 6`] = `0.8024691358024691`;

exports[`ease methods the easeOutQuad method should give the correct value 7`] = `0.8888888888888888`;

exports[`ease methods the easeOutQuad method should give the correct value 8`] = `0.9506172839506173`;

exports[`ease methods the easeOutQuad method should give the correct value 9`] = `0.9876543209876543`;

exports[`ease methods the easeOutQuad method should give the correct value 10`] = `1`;

exports[`ease methods the easeOutQuart method should give the correct value 1`] = `0`;

exports[`ease methods the easeOutQuart method should give the correct value 2`] = `0.375704923030026`;

exports[`ease methods the easeOutQuart method should give the correct value 3`] = `0.6340496875476299`;

exports[`ease methods the easeOutQuart method should give the correct value 4`] = `0.802469135802469`;

exports[`ease methods the easeOutQuart method should give the correct value 5`] = `0.9047401310775796`;

exports[`ease methods the easeOutQuart method should give the correct value 6`] = `0.9609815576893767`;

exports[`ease methods the easeOutQuart method should give the correct value 7`] = `0.9876543209876543`;

exports[`ease methods the easeOutQuart method should give the correct value 8`] = `0.997561347355586`;

exports[`ease methods the easeOutQuart method should give the correct value 9`] = `0.9998475842097241`;

exports[`ease methods the easeOutQuart method should give the correct value 10`] = `1`;

exports[`ease methods the easeOutQuint method should give the correct value 1`] = `0`;

exports[`ease methods the easeOutQuint method should give the correct value 2`] = `0.4450710426933565`;

exports[`ease methods the easeOutQuint method should give the correct value 3`] = `0.7153719792037121`;

exports[`ease methods the easeOutQuint method should give the correct value 4`] = `0.868312757201646`;

exports[`ease methods the easeOutQuint method should give the correct value 5`] = `0.9470778505986553`;

exports[`ease methods the easeOutQuint method should give the correct value 6`] = `0.9826584700841674`;

exports[`ease methods the easeOutQuint method should give the correct value 7`] = `0.9958847736625515`;

exports[`ease methods the easeOutQuint method should give the correct value 8`] = `0.9994580771901302`;

exports[`ease methods the easeOutQuint method should give the correct value 9`] = `0.9999830649121916`;

exports[`ease methods the easeOutQuint method should give the correct value 10`] = `1`;

exports[`ease methods the easeOutSine method should give the correct value 1`] = `0`;

exports[`ease methods the easeOutSine method should give the correct value 2`] = `0.17364817766693041`;

exports[`ease methods the easeOutSine method should give the correct value 3`] = `0.3420201433256689`;

exports[`ease methods the easeOutSine method should give the correct value 4`] = `0.4999999999999999`;

exports[`ease methods the easeOutSine method should give the correct value 5`] = `0.6427876096865394`;

exports[`ease methods the easeOutSine method should give the correct value 6`] = `0.766044443118978`;

exports[`ease methods the easeOutSine method should give the correct value 7`] = `0.8660254037844386`;

exports[`ease methods the easeOutSine method should give the correct value 8`] = `0.9396926207859084`;

exports[`ease methods the easeOutSine method should give the correct value 9`] = `0.984807753012208`;

exports[`ease methods the easeOutSine method should give the correct value 10`] = `1`;
